plonky3/src/circuit_builder.rs

```diff
@@ -323,3 +397,90 @@ where
         }
     }
 }
+
+impl<
+        T: FieldElementMap,
+        AB: AirBuilderWithPublicValues<F = Plonky3Field<T>> + PairBuilder + MultistageAirBuilder,
+    > MultiStageAir<AB> for PowdrCircuit<T>
+where
+    ProverData<T>: Send,
+    Commitment<T>: Send,
+{
+    fn stage_count(&self) -> usize {
+        self.constraint_system.stage_widths.len()
+    }
+
+    fn stage_trace_width(&self, stage: u32) -> usize {
+        self.constraint_system.stage_widths[stage as usize]
+    }
+
+    fn stage_challenge_count(&self, stage: u32) -> usize {
+        self.constraint_system
+            .challenges
+            .iter()
+            .filter(|c| c.stage == stage)
+            .count()
+    }
+}
+
+impl<T: FieldElementMap> NextStageTraceCallback<T::Config> for PowdrCircuit<T>
+where
+    ProverData<T>: Send,
+    Commitment<T>: Send,
+{
+    // this wraps the witgen callback to make it compatible with p3:
+    // - p3 passes its local challenge values and the stage id
+    // - it receives the trace for the next stage in the expected format, as well as the public values for this stage and the updated challenges for the previous stage
+    // - internally, the full witness is accumulated in [Self] as it's needed in order to call the witgen callback
```

Comment:
```suggestion
    // - it receives the trace for the next stage in the expected format,
    //   as well as the public values for this stage and the updated challenges
    //   for the previous stage
    // - internally, the full witness is accumulated in [Self] as it's needed
    //   in order to call the witgen callback
```
Just some line breaks